### ISSUE TITLE
acme.sh Please specify at least one validation method.

### ISSUE THREAD

**Issue by makuser** [Froxlor/Froxlor#816]
# Summary

SSL certs do not update since one of the last updates.

# System information
* Froxlor version: 0.10.15 (DB: 202002290), but also on 0.10.12.
* Web server: apache2:amd64/stable 2.4.38-3+deb10u3
* DNS server: bind9:amd64/stable 1:9.11.5.P4+dfsg-5.1
* POP/IMAP server: not on this server
* SMTP server: not on this server
* FTP server: ---
* OS/Version: Debian GNU/Linux 10 (buster)

# Steps to reproduce

1. Enable SSL.

# Expected behavior

1. Certificate updates

# Actual behavior

1. Certificate does not update

# Log files/log entries
<pre>
[information] Requesting/renewing Let's Encrypt certificates
[information] Updating certificate for domain.tld
[information] Adding SAN entry: domain.tld
[information] Adding SAN entry: www.domain.tld
[information] Checking for LetsEncrypt client upgrades before renewing certificates:
[Sun 08 Mar 2020 07:55:10 PM CET] Already uptodate!
[Sun 08 Mar 2020 07:55:10 PM CET] Upgrade success!
[Sun 08 Mar 2020 07:55:10 PM CET] Removing cron job
[Sun 08 Mar 2020 07:55:10 PM CET] Lets find script dir.
[Sun 08 Mar 2020 07:55:10 PM CET] _SCRIPT_='/root/.acme.sh/acme.sh'
[Sun 08 Mar 2020 07:55:10 PM CET] _script='/root/.acme.sh/acme.sh'
[Sun 08 Mar 2020 07:55:10 PM CET] _script_home='/root/.acme.sh'
[Sun 08 Mar 2020 07:55:10 PM CET] Using config home:/root/.acme.sh
[Sun 08 Mar 2020 07:55:10 PM CET] Using server: https://acme-v02.api.letsencrypt.org/directory
[Sun 08 Mar 2020 07:55:10 PM CET] Running cmd: renew
[Sun 08 Mar 2020 07:55:10 PM CET] Using config home:/root/.acme.sh
[Sun 08 Mar 2020 07:55:10 PM CET] ACME_DIRECTORY='https://acme-v02.api.letsencrypt.org/directory'
[Sun 08 Mar 2020 07:55:10 PM CET] DOMAIN_PATH='/root/.acme.sh/domain.tld'
[Sun 08 Mar 2020 07:55:10 PM CET] Le_API
Please specify at least one validation method: '--webroot', '--standalone', '--apache', '--nginx' or '--dns' etc.
[debug] https://github.com/acmesh-official/acme.sh
v2.8.6
[Sun 08 Mar 2020 07:55:10 PM CET] Renew: 'domain.tld'
[information] Updated Let's Encrypt certificate for domain.tld
[information] Updating certificate for domain.tld
[information] Adding SAN entry: domain.tld
[information] Adding SAN entry: www.domain.tld
[Sun 08 Mar 2020 07:55:10 PM CET] Lets find script dir.
[Sun 08 Mar 2020 07:55:10 PM CET] _SCRIPT_='/root/.acme.sh/acme.sh'
[Sun 08 Mar 2020 07:55:10 PM CET] _script='/root/.acme.sh/acme.sh'
[Sun 08 Mar 2020 07:55:10 PM CET] _script_home='/root/.acme.sh'
[Sun 08 Mar 2020 07:55:10 PM CET] Using config home:/root/.acme.sh
[Sun 08 Mar 2020 07:55:10 PM CET] Using server: https://acme-v02.api.letsencrypt.org/directory
[Sun 08 Mar 2020 07:55:10 PM CET] Running cmd: renew
[Sun 08 Mar 2020 07:55:10 PM CET] Using config home:/root/.acme.sh
[Sun 08 Mar 2020 07:55:10 PM CET] ACME_DIRECTORY='https://acme-v02.api.letsencrypt.org/directory'
[Sun 08 Mar 2020 07:55:10 PM CET] DOMAIN_PATH='/root/.acme.sh/domain.tld'
[Sun 08 Mar 2020 07:55:10 PM CET] Le_API
Please specify at least one validation method: '--webroot', '--standalone', '--apache', '--nginx' or '--dns' etc.
[debug] https://github.com/acmesh-official/acme.sh
v2.8.6
[Sun 08 Mar 2020 07:55:10 PM CET] Renew: 'domain.tld'
[information] Updated Let's Encrypt certificate for domain.tld
</pre>

**Comment by d00p** [comment]
your log says "updating certificate" ....there is not validation method to specify for renew, see https://github.com/acmesh-official/acme.sh

**Comment by d00p** [comment]
Also: works fine for me

**Comment by makuser** [comment]
> your log says "updating certificate" ....there is not validation method to specify for renew, see https://github.com/acmesh-official/acme.sh

Maybe I overlooked it, but where in Froxlor do I set that again?

**Comment by d00p** [comment]
what do you mean "where do I set that"? Set what?

**Comment by makuser** [comment]
> what do you mean "where do I set that"? Set what?

Where do I set a validation method? Because apparently acme.sh refuses to update the certificate.

**Comment by d00p** [comment]
You dont. Froxlor does that automatically depending on either it's a new certificate (issue) or a renewal (renew)...try deleting the certificate in froxlor and via acme.sh directly, then run the cronjob again. It should get issued. Else please use the forum for support, I cannot reproduce that

**Comment by makuser** [comment]
> Also: works fine for me

:rofl:

**Comment by makuser** [comment]
> Else please use the forum for support, I cannot reproduce that

I am not asking for support. I am trying to report a bug, that I encountered on both of my installations.

I also have the problem that a www subdomain gets added to the acme request. Who is at fault there? acme.sh or froxlor?

```
[information] Updating certificate for sub.domain.tld
[information] Adding SAN entry: sub.domain.tld
[Sun 08 Mar 2020 08:55:25 PM CET] Lets find script dir.
[Sun 08 Mar 2020 08:55:25 PM CET] _SCRIPT_='/root/.acme.sh/acme.sh'
[Sun 08 Mar 2020 08:55:25 PM CET] _script='/root/.acme.sh/acme.sh'
[Sun 08 Mar 2020 08:55:25 PM CET] _script_home='/root/.acme.sh'
[Sun 08 Mar 2020 08:55:25 PM CET] Using config home:/root/.acme.sh
[Sun 08 Mar 2020 08:55:25 PM CET] Using server: https://acme-v02.api.letsencrypt.org/directory
[Sun 08 Mar 2020 08:55:25 PM CET] Running cmd: renew
[Sun 08 Mar 2020 08:55:25 PM CET] Using config home:/root/.acme.sh
[Sun 08 Mar 2020 08:55:25 PM CET] ACME_DIRECTORY='https://acme-v02.api.letsencrypt.org/directory'
[Sun 08 Mar 2020 08:55:25 PM CET] DOMAIN_PATH='/root/.acme.sh/sub.domain.tld'
[Sun 08 Mar 2020 08:55:25 PM CET] Le_API
[Sun 08 Mar 2020 08:55:25 PM CET] _main_domain='sub.domain.tld'
[Sun 08 Mar 2020 08:55:25 PM CET] _alt_domains='www.sub.domain.tld'
[Sun 08 Mar 2020 08:55:25 PM CET] Using ACME_DIRECTORY: https://acme-v02.api.letsencrypt.org/directory
[Sun 08 Mar 2020 08:55:25 PM CET] _init api for server: https://acme-v02.api.letsencrypt.org/directory
[Sun 08 Mar 2020 08:55:25 PM CET] GET
[Sun 08 Mar 2020 08:55:25 PM CET] url='https://acme-v02.api.letsencrypt.org/directory'
[Sun 08 Mar 2020 08:55:25 PM CET] timeout=
[Sun 08 Mar 2020 08:55:25 PM CET] _CURL='curl -L --silent --dump-header /root/.acme.sh/http.header  -g '
[Sun 08 Mar 2020 08:55:25 PM CET] ret='0'
[Sun 08 Mar 2020 08:55:25 PM CET] ACME_KEY_CHANGE='https://acme-v02.api.letsencrypt.org/acme/key-change'
[Sun 08 Mar 2020 08:55:25 PM CET] ACME_NEW_AUTHZ
[Sun 08 Mar 2020 08:55:25 PM CET] ACME_NEW_ORDER='https://acme-v02.api.letsencrypt.org/acme/new-order'
[Sun 08 Mar 2020 08:55:25 PM CET] ACME_NEW_ACCOUNT='https://acme-v02.api.letsencrypt.org/acme/new-acct'
[Sun 08 Mar 2020 08:55:25 PM CET] ACME_REVOKE_CERT='https://acme-v02.api.letsencrypt.org/acme/revoke-cert'
[Sun 08 Mar 2020 08:55:25 PM CET] ACME_AGREEMENT='https://letsencrypt.org/documents/LE-SA-v1.2-November-15-2017.pdf'
[Sun 08 Mar 2020 08:55:25 PM CET] ACME_NEW_NONCE='https://acme-v02.api.letsencrypt.org/acme/new-nonce'
[Sun 08 Mar 2020 08:55:25 PM CET] ACME_VERSION='2'
[Sun 08 Mar 2020 08:55:26 PM CET] Le_NextRenewTime='1583337513'
[Sun 08 Mar 2020 08:55:26 PM CET] _on_before_issue
[Sun 08 Mar 2020 08:55:26 PM CET] _chk_main_domain='sub.domain.tld'
[Sun 08 Mar 2020 08:55:26 PM CET] _chk_alt_domains='www.sub.domain.tld'
[Sun 08 Mar 2020 08:55:26 PM CET] Le_LocalAddress
[Sun 08 Mar 2020 08:55:26 PM CET] d='sub.domain.tld'
[Sun 08 Mar 2020 08:55:26 PM CET] Check for domain='sub.domain.tld'
[Sun 08 Mar 2020 08:55:26 PM CET] _currentRoot='/var/www/froxlor'
[Sun 08 Mar 2020 08:55:26 PM CET] d='www.sub.domain.tld'
[Sun 08 Mar 2020 08:55:26 PM CET] Check for domain='www.sub.domain.tld'
[Sun 08 Mar 2020 08:55:26 PM CET] _currentRoot='/var/www/froxlor'
[Sun 08 Mar 2020 08:55:26 PM CET] d
[Sun 08 Mar 2020 08:55:26 PM CET] _saved_account_key_hash is not changed, skip register account.
[Sun 08 Mar 2020 08:55:26 PM CET] Read key length:4096
[Sun 08 Mar 2020 08:55:26 PM CET] Using config home:/root/.acme.sh
[Sun 08 Mar 2020 08:55:26 PM CET] ACME_DIRECTORY='https://acme-v02.api.letsencrypt.org/directory'
[Sun 08 Mar 2020 08:55:26 PM CET] Use length 4096
[Sun 08 Mar 2020 08:55:26 PM CET] Using RSA: 4096
[Sun 08 Mar 2020 08:55:27 PM CET] _createcsr
[Sun 08 Mar 2020 08:55:27 PM CET] d='www.sub.domain.tld'
[Sun 08 Mar 2020 08:55:27 PM CET] d
[Sun 08 Mar 2020 08:55:27 PM CET] url='https://acme-v02.api.letsencrypt.org/acme/new-order'
[Sun 08 Mar 2020 08:55:27 PM CET] payload='{"identifiers": [{"type":"dns","value":"sub.domain.tld"},{"type":"dns","value":"www.sub.domain.tld"}]}'
[Sun 08 Mar 2020 08:55:27 PM CET] RSA key
[Sun 08 Mar 2020 08:55:27 PM CET] HEAD
[Sun 08 Mar 2020 08:55:27 PM CET] _post_url='https://acme-v02.api.letsencrypt.org/acme/new-nonce'
[Sun 08 Mar 2020 08:55:27 PM CET] _CURL='curl -L --silent --dump-header /root/.acme.sh/http.header  -g  -I  '
[Sun 08 Mar 2020 08:55:28 PM CET] _ret='0'
[Sun 08 Mar 2020 08:55:28 PM CET] POST
[Sun 08 Mar 2020 08:55:28 PM CET] _post_url='https://acme-v02.api.letsencrypt.org/acme/new-order'
[Sun 08 Mar 2020 08:55:28 PM CET] _CURL='curl -L --silent --dump-header /root/.acme.sh/http.header  -g '
[Sun 08 Mar 2020 08:55:29 PM CET] _ret='0'
[Sun 08 Mar 2020 08:55:29 PM CET] code='201'
[Sun 08 Mar 2020 08:55:29 PM CET] Le_LinkOrder='https://acme-v02.api.letsencrypt.org/acme/order/70501337/2583991337'
[Sun 08 Mar 2020 08:55:29 PM CET] Le_OrderFinalize='https://acme-v02.api.letsencrypt.org/acme/finalize/70501337/2583991337'
[Sun 08 Mar 2020 08:55:29 PM CET] url='https://acme-v02.api.letsencrypt.org/acme/authz-v3/3177861337'
[Sun 08 Mar 2020 08:55:29 PM CET] payload
[Sun 08 Mar 2020 08:55:29 PM CET] POST
[Sun 08 Mar 2020 08:55:29 PM CET] _post_url='https://acme-v02.api.letsencrypt.org/acme/authz-v3/3177861337'
[Sun 08 Mar 2020 08:55:29 PM CET] _CURL='curl -L --silent --dump-header /root/.acme.sh/http.header  -g '
[Sun 08 Mar 2020 08:55:29 PM CET] _ret='0'
[Sun 08 Mar 2020 08:55:29 PM CET] code='200'
[Sun 08 Mar 2020 08:55:29 PM CET] url='https://acme-v02.api.letsencrypt.org/acme/authz-v3/3242091337'
[Sun 08 Mar 2020 08:55:29 PM CET] payload
[Sun 08 Mar 2020 08:55:29 PM CET] POST
[Sun 08 Mar 2020 08:55:29 PM CET] _post_url='https://acme-v02.api.letsencrypt.org/acme/authz-v3/3242091337'
[Sun 08 Mar 2020 08:55:29 PM CET] _CURL='curl -L --silent --dump-header /root/.acme.sh/http.header  -g '
[Sun 08 Mar 2020 08:55:30 PM CET] _ret='0'
[Sun 08 Mar 2020 08:55:30 PM CET] code='200'
[Sun 08 Mar 2020 08:55:30 PM CET] d='sub.domain.tld'
[Sun 08 Mar 2020 08:55:30 PM CET] _w='/var/www/froxlor'
[Sun 08 Mar 2020 08:55:30 PM CET] _currentRoot='/var/www/froxlor'
```

**Comment by d00p** [comment]
> I also have the problem that a www subdomain gets added to the acme request. Who is at fault 
> there? acme.sh or froxlor?

I'd say you've specified "www-alias" in the "ServerAlias value" setting for the domain.

And sorry, I've JUST added a new subdomain and set it to use let's encrypt and it worked just fine. I sadly do not have anything near a renew currently, if you are sure, please specify more details about domains configuration, dns entries, etc.

**Comment by d00p** [comment]
So, how convenient, froxlor.org's certificate just renewed on 6th of march so at least it's definetly not 0.10.12 as you've said (at most 0.10.14 as it was released after the renew). And as you can see here: https://github.com/Froxlor/Froxlor/compare/0.10.13...0.10.15 nothing regarding let's encrypt / acme.sh-implementation changed since then

**Comment by makuser** [comment]
> > I also have the problem that a www subdomain gets added to the acme request. Who is at fault
> > there? acme.sh or froxlor?
> 
> I'd say you've specified "www-alias" in the "ServerAlias value" setting for the domain.

No, I checked that already. I might have in the past for this specific domain, but this is long undone and set to none/0 in the db.

> And sorry, I've JUST added a new subdomain and set it to use let's encrypt and it worked just fine. I sadly do not have anything near a renew currently, if you are sure, please specify more details about domains configuration, dns entries, etc.

Yes, I have literally hundreds of other domains where its fine, but there are five certificates where it fails. One of them has an added www. SAN that is not specified in froxlor. I just edited the acme.sh conf file for this specific domain, saw a Le_Alt entry, where none should have been, so I deleted that line and rerun the froxlor --letsencrypt --debug command. Then this certificate updated just fine.

Why was the configuration file not changed? It looks like it contained old information, or at least information that is definitely incorrect at the time of usage.

**Comment by makuser** [comment]
> So, how convenient, froxlor.org's certificate just renewed on 6th of march so at least it's definetly not 0.10.12 as you've said (at most 0.10.14 as it was released after the renew). And as you can see here: [0.10.13...0.10.15](https://github.com/Froxlor/Froxlor/compare/0.10.13...0.10.15) nothing regarding let's encrypt / acme.sh-implementation changed since then

I only said it also did not work in 0.10.12. This is a fact.

The problem might be before 0.10.12 though, and because I did not use every version before that, I can not say exactly which one introduced it. The one I used are listed here:
```
-rw-r--r-- 1 froxlorlocal froxlorlocal 1605600 Mar  8 07:45 froxlor-0.10.15.tar.gz
-rw-r--r-- 1 froxlorlocal froxlorlocal 1603054 Jan 10 17:51 froxlor-0.10.12.tar.gz
-rw-r--r-- 1 froxlorlocal froxlorlocal 1598862 Dec 13 17:09 froxlor-0.10.10.tar.gz
-rw-r--r-- 1 froxlorlocal froxlorlocal 1586611 Oct 25 15:01 froxlor-0.10.3.tar.gz
-rw-r--r-- 1 froxlorlocal froxlorlocal 1273381 Nov 27  2018 froxlor-0.9.40.1.tar.gz
-rw-r--r-- 1 froxlorlocal froxlorlocal 1278129 Feb 14  2018 froxlor-0.9.39.5.tar.gz
-rw-r--r-- 1 froxlorlocal froxlorlocal 1212479 Dec 31  2017 froxlor-0.9.38.8.tar.gz
-rw-r--r-- 1 froxlorlocal froxlorlocal 1202009 Jan 24  2017 froxlor-0.9.38.7.tar.gz
-rw-r--r-- 1 froxlorlocal froxlorlocal 1202141 Nov 27  2016 froxlor-0.9.38.4.tar.gz
```
But I unfortunately also don't know, if 0.10.3 worked... I only know for sure, that 0.9.40.1 and every version before that worked 100%, but as far as I remember, they also used a different method (no acme.sh, right?), so that might not be helpful either.

**Comment by d00p** [comment]
> No, I checked that already. I might have in the past for this specific domain, but this is long undone and set to none/0 in the db.

Manual edits in the database are not intended. That way froxlor does not know about these changes and does not trigger these changes.

> I only said it also did not work in 0.10.12. This is a fact.

Did for me. Fact.

0.9.x used a totally different lets encrypt implementation, you are right. Again, without more details about domain settings, dns settings for that domain, possible existing files, etc. I can only say: this works for me on more than two servers

**Comment by makuser** [comment]
> > No, I checked that already. I might have in the past for this specific domain, but this is long undone and set to none/0 in the db.
> 
> Manual edits in the database are not intended. That way froxlor does not know about these changes and does not trigger these changes.

Yes, I know about that too, but these changes must have been done via the UI, because my clients do not have access to the database.

> > I only said it also did not work in 0.10.12. This is a fact.
> 
> Did for me. Fact.

That is great for you, and as I said the majority of my domains works just fine too. I just want to remark again, that this problem which for some reason prevented the certificate update already existed in 0.10.12 and because lack of time to investigate, I had to generate the certificate externally, and use the custom certificate option in Froxlor to supply the generated certificate. I thought an update would fix it, but even the latest 0.10.15 that I installed yesterday did not solve this.

> 0.9.x used a totally different lets encrypt implementation, you are right. Again, without more details about domain settings, dns settings for that domain, possible existing files, etc. I can only say: this works for me on more than two servers

Well, I'm still trying to figure out what is different about those domains that do not work. I yet have no clue at all, because I can not find any differences to domains that work, which leads me to believe, that even those domains do not work due to some prior settings that have been changed since then.
